### PR TITLE
Add support for OAuth 2.0 (Gmail)

### LINKED ISSUE #655: Add support for Gmail's XOAUTH2
I really do not want to allow "less secure apps" in my gmail account.
Would K9Mail be compatible with this gmail security some time in a future?

### PR DESCRIPTION
**Pull request by cketti**
The implementation cuts a few corners. The plan is to clean this up when we rework the account setup code to support Thunderbird's Autoconfiguration (#865).

Fixes #655

### CODE REVIEW

**Comment by Monkey-Matt** [comment]
Been testing this out and looks to be working well.

One (minor) unfortunate case I found was if you enter a non gmail email address (eg spelling mistake) and click next it shows the password field, then if you change the email to a gmail account it still requires the password to be filled before you can continue. If you enter something in the password it does navigate through the oauth flow correctly so it's not much of an issue.

If this progresses is it possibly to "verify the app with google" so we don't get the aggressive warning? Not sure what that involves but I'm just imagining the concerned forum posts that might result on seeing the warning.

**Comment by mmhere** [comment]
What's the best way to take advantage of the hard work accomplished here in enabling OAuth?  It looks like the [much appreciated] effort is merged into main, but I've only ever installed .apk files from https://github.com/k9mail/k-9/releases.

Google's "30 May" deadline looms (as I'm sure everyone is aware, but I thought I'd bang the drum one more time anyway -- please don't take that the wrong way as it is only meant as a reminder).

**Comment by Monkey-Matt** [comment]
> What's the best way to take advantage of the hard work accomplished here in enabling OAuth? It looks like the [much appreciated] effort is merged into main, but I've only ever installed .apk files from https://github.com/k9mail/k-9/releases.

Until there is a new release the only way to install the latest work is to build the project yourself.

> Google's "30 May" deadline looms (as I'm sure everyone is aware, but I thought I'd bang the drum one more time anyway -- please don't take that the wrong way as it is only meant as a reminder).

Existing/previous versions of K-9 will continue to work with gmail after that deadline, you simply have to use an "app password" from gmail. https://forum.k9mail.app/t/oauth2-authentication/2671/38?u=monkeymatt

**Comment by cketti** [comment]
See https://forum.k9mail.app/t/the-plan-for-k-9-mail-6-200/4590/4

**Comment by mmhere** [comment]
comments in response to "just use 2FA for app password" over there:https://forum.k9mail.app/t/the-plan-for-k-9-mail-6-200/4590/5  so as not to turn this here into a discussion thread.

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/setup/AuthViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    private fun createAuthorizationRequestIntent(email: String, config: OAuthConfiguration): Intent {
+        val serviceConfig = AuthorizationServiceConfiguration(
+            config.authorizationEndpoint.toUri(),
+            config.tokenEndpoint.toUri()
+        )
+
+        val applicationId = getApplication<Application>().packageName
+        val redirectUri = Uri.parse("$applicationId:/oauth2redirect")
```
Comment (blino) [inline app/ui/legacy/src/main/java/com/fsck/k9/activity/setup/AuthViewModel.kt:114]: Shouldn't this be `$applicationId://oauth2redirect` with two slashes?
It seems the Microsoft Azure console does not allow adding a redirection URI with just one slash.

**Comment by drago333** [comment]
> Is this implementation planning to support custom domains that use Google Workspace?

I tested it with my school email which is on the school domain and it works.

File: app/ui/legacy/src/main/java/com/fsck/k9/activity/setup/AuthViewModel.kt (hunk 1 above)
Comment (blino) [reply app/ui/legacy/src/main/java/com/fsck/k9/activity/setup/AuthViewModel.kt:114]: I am trying to make this work with Office365/Outlook.
Are you maybe already working on this?
If not, I will open a new issue/MR for it.
But the blocker point will likely be to setup a k9mail application on Microsoft Azure, to have a dedicated client-id and allow the k9mail redirection URL.

**Comment by x2611** [comment]
> However, if you manually set up the account and select "OAuth 2.0" under "Authentication", it should work as long as the server names are `imap.gmail.com` and `smtp.gmail.com`.

There is no "OAuth 2.0" under "Authentication".
Only:
Normal password
Encrypted password
Client certificate

v6.000. What am I missing?

**Pull request merged**